### PR TITLE
docx and FB3 formats support (for QT only)

### PR DESCRIPTION
**Pull request by pkb**
### FB3 support
- See #105 for details
### WordML (docx) support
The following features are supported

- Basic formatting (bold,italic,underline,strike-through,subscript,superscript)
- Paragraph alignment
- Images (formats supported by crengine. i.e. no emf,wmf)
- Partially lists and tables
- Links, footnotes and endnotes

### CODE REVIEW

**Comment by virxkane** [comment]
Compilation failed with gcc-8.3.0 (Linux, Qt5).
```[ 55%] Building CXX object crengine/CMakeFiles/crengine.dir/src/docxfmt.cpp.o
cd /var/tmp/2/coolreader/debug-build/crengine && /usr/bin/c++  -DBIG_PAGE_MARGINS=0 -DCHM_SUPPORT_ENABLED=1 -DCOLOR_BACKBUFFER=1 -DCR3_ANTIWORD_PATCH=1 -DDEBUG=1 -DDOC_BUFFER_SIZE=0x400000 -DDOC_DATA_COMPRESSION_LEVEL=1 -DENABLE_ANTIWORD=1 -DLDOM_USE_OWN_MEM_MAN=1 -DLINUX=1 -DMAX_IMAGE_SCALE_MUL=0 -DQT_CORE_LIB -DQT_GUI_LIB -DQT_WIDGETS_LIB -DUSE_EXTERNAL_EDICT_DICTIONARY=0 -DUSE_FONTCONFIG=1 -DUSE_FREETYPE=1 -DUSE_UNRAR=0 -D_DEBUG=1 -D_LINUX=1 -I/usr/include/qt5 -I/usr/include/qt5/QtWidgets -I/usr/include/qt5/QtGui -I/usr/include/qt5/QtCore -I/usr/lib64/qt5/mkspecs/linux-g++ -I/var/tmp/2/coolreader/thirdparty/antiword -I/usr/include/freetype2 -I/usr/include/harfbuzz -I/var/tmp/2/coolreader/thirdparty/chmlib/src -I/var/tmp/2/coolreader/crengine/include -I/var/tmp/2/coolreader/tinydict -I/var/tmp/2/coolreader/crengine/fc-lang  -Wall -Wno-long-long -pedantic -std=c++11 -fPIC -Wno-misleading-indentation -g   -o CMakeFiles/crengine.dir/src/docxfmt.cpp.o -c /var/tmp/2/coolreader/crengine/src/docxfmt.cpp
/var/tmp/2/coolreader/crengine/src/docxfmt.cpp:353:14: error: explicit specialization in non-namespace scope 'class docx_PropertiesContainer<N>'
     template<>
              ^
/var/tmp/2/coolreader/crengine/src/docxfmt.cpp: In constructor 'docxImportContext::docxImportContext(OpcPackage*, ldomDocument*)':
/var/tmp/2/coolreader/crengine/src/docxfmt.cpp:657:16: warning: 'docxImportContext::m_pStyle' will be initialized after [-Wreorder]
     docxStyle* m_pStyle;
                ^~~~~~~~
/var/tmp/2/coolreader/crengine/src/docxfmt.cpp:620:17: warning:   'OpcPackage* docxImportContext::m_package' [-Wreorder]
     OpcPackage* m_package;
                 ^~~~~~~~~
/var/tmp/2/coolreader/crengine/src/docxfmt.cpp:2389:1: warning:   when initialized here [-Wreorder]
 docxImportContext::docxImportContext(OpcPackage *package, ldomDocument *doc) : m_styles(64), m_abstractNumbers(16),
 ^~~~~~~~~~~~~~~~~
make[2]: *** [crengine/CMakeFiles/crengine.dir/build.make:518: crengine/CMakeFiles/crengine.dir/src/docxfmt.cpp.o] Error 1
make[2]: Leaving directory '/var/tmp/2/coolreader/debug-build'
make[1]: *** [CMakeFiles/Makefile2:256: crengine/CMakeFiles/crengine.dir/all] Error 2
make[1]: Leaving directory '/var/tmp/2/coolreader/debug-build'
make: *** [Makefile:130: all] Error 2

**Comment by pkb** [comment]
@virxkane thanks, I will fix it today.

**Comment by virxkane** [comment]
1. Test file [sample3.docx](https://github.com/buggins/coolreader/files/3672635/sample3.docx)
In LibreOffice: sample3_lo
In CoolReader: sample3_cr3

1.1. Extraneous characters "1".
1.2. The text after the title is not wrapped to the next page, but perhaps this is not necessary.

**Comment by virxkane** [comment]
2. [Отрывок3.docx](https://github.com/buggins/coolreader/files/3672708/3.docx)
In LibreOffice: Отрывок3_lo
In CoolReader: Отрывок3_cr3
Extra underline.

**Comment by virxkane** [comment]
@pkb Sorry for the comment, but if you fix this it will be very cool.
P.S. Added into android layer.
https://github.com/virxkane/coolreader/tree/pkb_fb3_support
Screenshot_1569868650

**Comment by pkb** [comment]
> Extraneous characters "1" 

This one should be fixed. The underline issue is also clear but I don't have time to fix it right now. Will fix it later.

**Comment by pkb** [comment]
@virxkane thanks for testing

**Comment by pkb** [comment]
> Extra underline.

@virxkane, fixed.

**Comment by poire-z** [comment]
I cherry picked this PR into my KOReader crengine fork, and it's working quite well. All my docx test documents opened correctly, and are quite readable (some fonts and background colors missing here and there, but that's ok) - and they opened faster than with MS Word :)
It's quite cleaned and contained in dedicated new source files, so I expect no side effect on current code. But it's quite complex :) @pkb is that all your original work that you started from scratch ?! It's amazing :)
Are you ok if we pick this into KOReader's crengine? (and possibly merge it there before it's merged here?)

**Comment by pkb** [comment]
@poire-z, about colors and fonts, it is expected. I do not read this info. It could be added later as an option if there will be demand. Yes I implemented it from scratch. Sure you can pick it. Who knows when @buggins  will merge it. Why should you wait. :)

File: cr3qt/data/docx.css [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+a[type="note"] { $footnote-link.all }
+
+image { text-align: center; text-indent: 0px; display: block }
+p image { display: inline }
+li image { display: inline }
+
+li { display: list-item; text-indent: 0em;  }
+ol { display: block; list-style-type: decimal; margin-left: 1em }
+
+title.h1 p, title.h2 h2 p {
```
Comment (poire-z) [inline cr3qt/data/docx.css]: This additional `h2` between `title.h2` and `p` looks like a typo.

**Comment by poire-z** [comment]
Thanks.

I need to update this docx.css because we don't do the `$def` and `$title` stuff - It looks like you started this docx.css from scratch, and all these are needed (there's no remnants from other DTDs that I could be tempted to clean :).

The choice of building a DOM with `<title class=H3><p>The H3 title</p></title>` - and not a simpler `<h3>The H3 title</h3>` - is that your (arbitrary?) decision ? Or is it dictated by the DOCX own format/elements hierarchy?

Also, is my understanding right that there is no embedded CSS (files or inline `style=`) involved with your implementation? Because you didn't bother supporting that (yet :)? Or because there is no support for that (classic CSS) in DOCX?

So, we get a quite simple "rich text format", with section, title, p, lists, tables, and only 6 inline styles (strong, emphasis, u, strike, sup, sub), right ? (fine enough for me - I just want to understand how it goes :).

Also, we do have a different `css_length_t`, where we have the value `*256`, for higher precisions.
In docxfmt.cpp, there are some `css_length_t` with types `css_val_pt` and `css_val_in`.
It looks to me, none of these are real lengths, and you use them to store counters/IDs (`css_val_in` for some numberings/levels IDs, `css_val_pt` for some properties IDs - 'in' sounds like 'id', 'pt' sounds like 'properties' :). Is my understanding right?
So, it looks like there is currently no support for sizes, font sizes, margins, padding, line spacing? So I don't need to worry about these css_length_t. Right ? (except of course for those that we define in docx.css, but these are applied and handled outside docxfmt.cpp).

File: cr3qt/data/docx.css (hunk 1 above)
Comment (pkb) [reply cr3qt/data/docx.css]: Yes, it was typo

**Comment by pkb** [comment]
> is that your (arbitrary?) decision ? Or is it dictated by the DOCX own format/elements hierarchy?

It is my decision. I wanted to have TOC that crengine builds using title elements and I wanted to have simple way to configure appearance for different levels.

> Or because there is no support for that (classic CSS) in DOCX?

CSS is not used in DOCX. 

> So, we get a quite simple "rich text format",

Yes

> Is my understanding right?

Right :)

> So I don't need to worry about these css_length_t. Right ?

I was thinking about using it for sizes, font sizes, margins, padding, line spacing but you are right these are not supported so you don't need to worry :)

**Comment by poire-z** [comment]
Thank you for your answers.
OK, [PR](https://github.com/koreader/crengine/pull/314) made to KOReader's crengine. I made a few tweaks to adapt it to how we prefer stuff (no fb3.css, different handling of headings and footnotes). If you can have a quick look, and tell me if I did miss/mess some essential stuff with that section/title stuff (which I didn't really get :).

File: cr3qt/data/docx.css [hunk 2]
```diff
@@ -14,45 +14,45 @@ li image { display: inline }
 li { display: list-item; text-indent: 0em;  }
 ol { display: block; list-style-type: decimal; margin-left: 1em }
 
-title.h1 p, title.h2 p { 
+.h1, .h2 {
     $title.all
 }
 
-title.h3 p, title.h4 p, title.h5 p, title.h6 p { 
+.h3, .h4, .h5, .h6 p {
```
Comment (poire-z) [inline cr3qt/data/docx.css]: Probably typo in these (you removed the "p" in the "hx p" descendant rules. Either that's what you wanted and you forgot to remove it in the `h6 p`, or it's not and you removed too many (with `h3, h4, h5, h6 p`, you get 4 selectors: `h3` `h4` `h5` `h6 p` - it does _not_ become `h3 p` `h4 p` `h5 p` `h6 p`)

Comment (pkb) [reply cr3qt/data/docx.css]: Thanks @poire-z, indeed I forgot remove it in the h6 p

**Pull request merged**